File: S/1022-sum-of-root-to-leaf-binary-numbers/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sumRootToLeaf(root *TreeNode) int {
	var recur func(root *TreeNode, hier int, sum *int)
	recur = func(root *TreeNode, hier int, sum *int) {
		if root == nil {
			return
		}
		if root.Left == nil && root.Right == nil {
			*sum += hier*2 + root.Val
			return
		}
		recur(root.Left, hier*2+root.Val, sum)
		recur(root.Right, hier*2+root.Val, sum)
	}
	sum := 0
	recur(root, 0, &sum)
	return sum
}

func main() {
	fmt.Println("a")
}

File: S/1022-sum-of-root-to-leaf-binary-numbers/main.rs
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
  pub fn sum_root_to_leaf(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    fn recur(root: &Option<Rc<RefCell<TreeNode>>>, aggr: i32, ans: &mut i32) {
      match root {
        Some(root) => {
          let aggr = aggr * 2 + root.borrow().val;

          if let None = root.borrow().left
            && let None = root.borrow().right
          {
            *ans += aggr;
          } else {
            recur(&root.borrow().left, aggr, ans);
            recur(&root.borrow().right, aggr, ans);
          }
        }
        _ => (),
      }
    }

    let mut ans: i32 = 0;
    recur(&root, 0, &mut ans);
    ans
  }
}

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None,
    }
  }
}
